docs/http-cloud-provider.md

```diff
@@ -0,0 +1,643 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<strong>
+The latest 1.0.x release of this document can be found
+[here](http://releases.k8s.io/release-1.0/docs/http-cloud-provider.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# HTTP Cloudprovider
+
+# Motivation
+
+Currently, the cloudproviders are designed so that they get directly compiled into kubernetes. It is very hard to add custom cloudproviders to the kubernetes platform without using a custom build of kubernetes. Therefore, we would like to build an HTTP Cloudprovider which would be able to send a HTTP request by converting the cloudprovider interfaces to an HTTP API. The cloudprovider would need an HTTP endpoint so that kubernetes would be able to send a request to them and then the cloudproviders would be able to send the desired service information back to kubernetes.
+
+# Specification
+
+The user should specify a config file in a INI config-file format. The file is read using gcfg. The **client-url** field is required which specifies the URL of the cloud provider. The client-url field should come under the Global section.
+
+##### Example Config File:
+
+```bash
+[Global]
+client-url = http://127.0.0.1 
+```
+
+First the API checks for which interfaces in [cloud.go](../pkg/cloudprovider/cloud.go) are implemented.
+
+A GET request for /v1/{Interface}/supported will return whether the interface is supported or not.
+
+The following methods must be implemented by the HTTP cloudprovider. If a cloudprovider does not implement a resource, it can return false  or return a HTTP 404 Error with empty body for a given GET /{interface}/supported call.
+
+If you encounter an error in processing the request, return an HTTP 404 with a response body of:
+
+```json
+{
+    "error" :"error-message-here"
+}
+```
```

Comment:
Currently kubernetes api uses api.Status as the body of an error. Might consider reusing. 

https://github.com/kubernetes/kubernetes/blob/master/pkg/api/v1/types.go#L1586-L1607

If we want to reuse conversion mechanisms, this needs to at least have TypeMeta and ObjectMeta fields